mixer/adapter/svcctrl/checkprocessor.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcctrl
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	rpc "github.com/googleapis/googleapis/google/rpc"
+	"github.com/pborman/uuid"
+	sc "google.golang.org/api/servicecontrol/v1"
+
+	"istio.io/istio/mixer/adapter/svcctrl/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/apikey"
+)
+
+type checkProcImpl struct {
+	env                   adapter.Env
+	checkResponseCache    lruCache
+	checkResultExpiration time.Duration
+	consumerProjIDCache   lruCache
+	runtimeConfig         *config.RuntimeConfig
+	serviceConfig         *config.ServiceConfig
+	client                serviceControlClient
+}
+
+func (p *checkProcImpl) ProcessCheck(ctx context.Context, instance *apikey.Instance) (adapter.CheckResult, error) {
+	if instance.ApiKey == "" || instance.ApiOperation == "" {
+		return p.createCheckResult(rpc.Status{
+			Code:    int32(rpc.INVALID_ARGUMENT),
+			Message: "api key and api operation must not be empty",
+		}), nil
+	}
+
+	consumerID := apiKeyPrefix + instance.ApiKey
+	opName := instance.ApiOperation
+	response, err := p.checkImpl(consumerID, opName, instance.Timestamp)
+	if err != nil {
+		return p.createCheckResult(
+			rpc.Status{
+				Code:    int32(rpc.PERMISSION_DENIED),
+				Message: err.Error(),
+			}), nil
+	}
+
+	if p.env.Logger().VerbosityLevel(logDetail) {
+		if responseDetail, err := toFormattedJSON(response); err == nil {
+			p.env.Logger().Infof("response: %v", string(responseDetail))
+		}
+	}
+
+	return p.responseToCheckResult(response)
+}
+
+func (p *checkProcImpl) ResolveConsumerProjectID(consumerID, opName string) (string, error) {
+	consumerProjID, found := p.consumerProjIDCache.get(consumerID)
+	if !found {
+		_, err := p.checkImpl(consumerID, opName, time.Now())
+		if err != nil {
```

Comment:
Please comment todo: Mixer should provide a mechanism so that attributes returned by check can be used by a corresponding report.

This kind of adapter specific cache is not ideal.